store/postgres/migrations/2019-03-11-221621_delete_unneeded_entity_history/up.sql

```diff
@@ -0,0 +1,190 @@
+-- use from entity_history, so we don't get a conflict later when we
+-- drop that table
+alter table entity_history
+      alter column id drop default;
+alter sequence entity_history_id_seq
+      owned by eh.id;
+alter table eh drop column data_after;
+
+-- Lock any writes to entity_history and event_meta_data. We can not
+-- afford to have any changes happen while we copy, as we would lose
+-- them because we are copying
+lock table entity_history in exclusive mode;
+lock table event_meta_data in exclusive mode;
+
+-- Copy the good data. Do not log to avoid stress on the WAL. Note that
+-- either data_before or data_after can be null, and we need to make sure
+-- we include rows with a null in either column, but that 'anything !=
+-- null' will always be false. We lift SQL NULL to JSON null, since JSON
+-- uses normal, not three-state logic for comparisons with null
+alter table eh set unlogged;
+insert into eh
+  select id, event_id, entity_id, subgraph, entity, data_before,
+         reversion, op_id from entity_history
+  where subgraph != 'subgraphs'
+    and coalesce(data_before, 'null'::jsonb)
+        != coalesce(data_after, 'null'::jsonb);
+alter table eh set logged;
+
+-- Swap old and new tables
+alter table entity_history rename to eh_old;
+alter table eh rename to entity_history;
+
+-- Create indexes
+alter index entity_history_pkey rename to eh_old_pkey;
+alter table entity_history
+  add constraint entity_history_pkey primary key(id);
+
+alter index entity_history_event_id_btree_idx
+  rename to eh_old_event_id_btree_idx;
+create index entity_history_event_id_btree_idx
+  on entity_history(event_id);
+
+--
+-- Deal with event_meta_data
+--
+
+-- Remove uneeded event_meta_data entries following the same pattern
+create table emd(like event_meta_data
+                     including defaults
+                     including comments
+                     including storage);
+alter table event_meta_data
+      alter column id drop default;
+alter sequence event_meta_data_id_seq
+      owned by emd.id;
+
+-- Copy the good event_meta_data
+alter table emd set unlogged;
+insert into emd
+  select * from event_meta_data d
+  where exists (select 1 from entity_history h where d.id = h.event_id);
+alter table emd set logged;
+
+alter table event_meta_data rename to emd_old;
+alter table emd rename to event_meta_data;
+
+-- Create indexes
+alter index event_meta_data_pkey rename to emd_old_pkey;
+alter table event_meta_data
+  add constraint event_meta_data_pkey primary key(id);
+
+alter index event_meta_data_db_transaction_id_key
+  rename to emd_old_db_transaction_id_key;
+alter table event_meta_data
+  add constraint event_meta_data_db_transaction_id_key
+                 unique(db_transaction_id);
+
+-- Make sure the fk constraint on entity_history points to the new
+-- event_meta_data table
+alter table eh_old
+  drop constraint entity_history_event_id_fkey;
+
+alter table entity_history
```

Comment:
A comment here would be great; "match full on update cascade on delete cascade" is quite a mouthful. 😁

Reply:
That's just copy paste from `\d entity_history`; I'll add a comment.